clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java

```diff
@@ -0,0 +1,274 @@
+        NetworkClientDelegate.PollResult result = heartbeatRequestManager.poll(mockTime.milliseconds());
+        assertEquals(0, result.unsentRequests.size());
+
+        if (mockMembershipManager.shouldSendHeartbeat()) {
+            assertEquals(heartbeatInterval - 100, result.timeUntilNextPollMs);
+        } else {
+            assertEquals(Long.MAX_VALUE, result.timeUntilNextPollMs);
+        }
+    }
+
+    @Test
+    public void testNoCoordinator() {
+        when(mockCoordinatorRequestManager.coordinator()).thenReturn(Optional.empty());
+        NetworkClientDelegate.PollResult result = heartbeatRequestManager.poll(mockTime.milliseconds());
+
+        assertEquals(Long.MAX_VALUE, result.timeUntilNextPollMs);
+        assertEquals(0, result.unsentRequests.size());
+    }
+
+    @ParameterizedTest
+    @MethodSource("errorProvider")
+    public void testHeartbeatResponseOnErrorHandling(final Errors error, final boolean isFatal) {
+        heartbeatRequestState = new HeartbeatRequestManager.HeartbeatRequestState(
+            mockLogContext,
+            mockTime,
+            heartbeatInterval,
+            retryBackoffMs,
+            retryBackoffMaxMs,
+            0);
+        errorEventHandler = mock(ErrorEventHandler.class);
+        when(mockMembershipManager.state()).thenReturn(MemberState.RECONCILING);
+        heartbeatRequestManager = createManager();
+
+        mockTime.sleep(1001);
+        NetworkClientDelegate.PollResult result = heartbeatRequestManager.poll(mockTime.milliseconds());
+        assertEquals(1, result.unsentRequests.size());
+
+        // manually completing the response to test error handling
+        ClientResponse response = createHeartbeatResponse(
+            result.unsentRequests.get(0),
+            error);
+        result.unsentRequests.get(0).future().complete(response);
+        ConsumerGroupHeartbeatResponse mockResponse = (ConsumerGroupHeartbeatResponse) response.responseBody();
+
+        short errorCode = error.code();
+        if (errorCode == Errors.NONE.code()) {
+            verify(errorEventHandler, never()).handle(any());
+            verify(mockMembershipManager).updateState(mockResponse.data());
+            assertEquals(heartbeatInterval, heartbeatRequestState.nextHeartbeatMs(mockTime.milliseconds()));
+        } else if (isFatal) {
+            verify(errorEventHandler).handle(any());
+            assertEquals(retryBackoffMs, heartbeatRequestState.nextHeartbeatMs(mockTime.milliseconds()));
+            if (errorCode == Errors.UNRELEASED_INSTANCE_ID.code()) {
+                verify(mockMembershipManager).transitionToFailure();
+            } else if (errorCode == Errors.FENCED_MEMBER_EPOCH.code() ||
+                errorCode == Errors.UNKNOWN_MEMBER_ID.code()) {
+                verify(mockMembershipManager).fenceMember();
+            }
+            // The memberStateManager should have stopped heartbeat at this point
```

Comment:
From the HB request manager point of view, this means that when polled, it won't return any request right? Could we assert that to ensure that the request manager is actually not generating requests?